src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/Razor/RazorDynamicFileInfoProvider.TextChangesTextLoader.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.LanguageServer.LanguageServer;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.HostWorkspace.Razor;
+
+internal partial class RazorDynamicFileInfoProvider
+{
+    private sealed class TextChangesTextLoader(
+        TextDocument? document,
+        IEnumerable<TextChange> changes,
+        byte[] checksum,
+        SourceHashAlgorithm checksumAlgorithm,
+        int? codePage,
+        Uri razorUri) : TextLoader
+    {
+        private readonly Lazy<SourceText> _emptySourceText = new Lazy<SourceText>(() =>
+        {
+            var encoding = codePage is null ? null : Encoding.GetEncoding(codePage.Value);
+            return SourceText.From("", checksumAlgorithm: checksumAlgorithm, encoding: encoding);
+        });
+
+        public override async Task<TextAndVersion> LoadTextAndVersionAsync(LoadTextOptions options, CancellationToken cancellationToken)
+        {
+            if (document is null)
+            {
+                var text = _emptySourceText.Value.WithChanges(changes);
+                return TextAndVersion.Create(text, VersionStamp.Default.GetNewerVersion());
+            }
+
+            var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            // Validate the checksum information so the edits are known to be correct
+            if (IsSourceTextMatching(sourceText))
+            {
+                var version = await document.GetTextVersionAsync(cancellationToken).ConfigureAwait(false);
+                var newText = sourceText.WithChanges(changes);
+                return TextAndVersion.Create(newText, version.GetNewerVersion());
+            }
+
+            return await GetFullDocumentFromServerAsync(razorUri, cancellationToken).ConfigureAwait(false);
+        }
+
+        private bool IsSourceTextMatching(SourceText sourceText)
+        {
+            if (sourceText.ChecksumAlgorithm != checksumAlgorithm)
+            {
+                return false;
+            }
+
+            if (sourceText.Encoding?.CodePage == codePage)
```

Comment:
a couple things
1.  might be missing something - shouldn't this be `!=` (if the encodings are not equal we don't have a match)?
2.  i think the Encoding should override equals - https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Text/Encoding.cs#L1052
3.  the source text does include the encoding in the checksum - https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/Text/SourceTextStream.cs#L121

Reply:
nvm the other type isn't an `Encoding`, ignore 2

Reply:
hmm... I interpreted 3 slightly different: https://github.com/dotnet/roslyn/blob/ad6795a50273e90af90fa534498f856abed0b85e/src/Compilers/Core/Portable/Text/SourceText.cs#L584 

1. is definitely right. I'll add unit tests now that this seems the way forward.